src/units/unit.cpp

```diff
@@ -1442,116 +1442,6 @@ void unit::remove_ability_by_id(const std::string& ability)
-		if(filter["affect_adjacent"].to_bool() != adjacent){
-			return false;
-		}
-	}
-
-	if(!bool_matches_if_present(filter, cfg, "affect_self", true))
-		return false;
-
-	if(!bool_or_empty(filter, cfg, "affect_allies"))
-		return false;
-
-	if(!bool_matches_if_present(filter, cfg, "affect_enemies", false))
-		return false;
-
-	if(!bool_matches_if_present(filter, cfg, "cumulative", false))
-		return false;
-
-	const std::vector<std::string> filter_type = utils::split(filter["tag_name"]);
-	if ( !filter_type.empty() && std::find(filter_type.begin(), filter_type.end(), tag_name) == filter_type.end() )
-		return false;
-
-	if(!string_matches_if_present(filter, cfg, "id", ""))
-		return false;
-
-	if(!string_matches_if_present(filter, cfg, "apply_to", "self"))
-		return false;
-
-	if(!string_matches_if_present(filter, cfg, "overwrite_specials", "none"))
-		return false;
-
-	if(!string_matches_if_present(filter, cfg, "active_on", "both"))
-		return false;
-
-	//for damage only
-	if(!string_matches_if_present(filter, cfg, "replacement_type", ""))
-		return false;
-
-	if(!string_matches_if_present(filter, cfg, "alternative_type", ""))
-		return false;
-
-	//for plague only
-	if(!string_matches_if_present(filter, cfg, "type", ""))
-		return false;
-
-	if(!filter["value"].empty()){
-		if(tag_name == "drains"){
-			if(!int_matches_if_present(filter, cfg, "value", 50)){
-				return false;
-			}
-		} else if(tag_name == "berserk"){
-			if(!int_matches_if_present(filter, cfg, "value", 1)){
-				return false;
-			}
-		} else if(tag_name == "heal_on_hit" || tag_name == "heals" || tag_name == "regenerate" || tag_name == "leadership"){
-			if(!int_matches_if_present(filter, cfg, "value" , 0)){
-				return false;
-			}
-		} else {
-			if(!int_matches_if_present(filter, cfg, "value")){
-				return false;
-			}
-		}
-	}
-
-	if(!int_matches_if_present_or_negative(filter, cfg, "add", "sub"))
-		return false;
-
-	if(!int_matches_if_present_or_negative(filter, cfg, "sub", "add"))
-		return false;
-
-	if(!double_matches_if_present(filter, cfg, "multiply"))
-		return false;
-
-	if(!double_matches_if_present(filter, cfg, "divide"))
-		return false;
-
-	// Passed all tests.
-	return true;
-}
-
-bool unit::ability_matches_filter(const config & cfg, const std::string& tag_name, const config & filter) const
-{
-	// Handle the basic filter.
-	bool matches = matches_ability_filter(cfg, tag_name, filter);
-
-	// Handle [and], [or], and [not] with in-order precedence
-	for (const config::any_child condition : filter.all_children_range() )
-	{
-		// Handle [and]
-		if ( condition.key == "and" )
-			matches = matches && ability_matches_filter(cfg, tag_name, condition.cfg);
-
-		// Handle [or]
-		else if ( condition.key == "or" )
-			matches = matches || ability_matches_filter(cfg, tag_name, condition.cfg);
-
-		// Handle [not]
-		else if ( condition.key == "not" )
-			matches = matches && !ability_matches_filter(cfg, tag_name, condition.cfg);
-	}
-
-	return matches;
-}
-
```

Comment:
If the PR is just to add `[filter_wml]`, why is a large amount of code moved between files?

Reply:
the first commit here is the same what https://github.com/wesnoth/wesnoth/pull/8561 who don't be merged yet, and i don't want wait more longer for submit [filter_wml] and @soliton- refuse what i pull in https://github.com/wesnoth/wesnoth/pull/8561

Reply:
The idea is that a small PR, adding only `[filter_wml]`, could be merged before #8561.